### ISSUE TITLE
[BUG] wasmcloud does not seem to properly stop actors

### ISSUE THREAD

**Issue by jsoverson** [wasmCloud/wasmCloud#103]
**Describe the bug**

Wasmcloud seems to keep actors hanging around even after stopping them through wash

**To Reproduce**

With wasmcloud running somewhere:

```bash
# Convenient alias
$ alias get-inventory="wash ctl get hosts -o json | jq -r '.hosts[].id' | xargs -n 1 wash ctl get inventory -o json | jq -s 'map({host:.inventory.host_id, actors: (.inventory.actors)})'"
# Show no actors running on hosts
$ get-inventory
# start arbitrary actor
$ wash ctl start actor wasmcloud.azurecr.io/echo:0.2.0
# show actor is running
$ get-inventory
# get actor id to reference later
$ ACTOR_ID=$(wash claims inspect -o json wasmcloud.azurecr.io/echo:0.2.0 | jq -r '.module')
# make a call to the actor 
$ wash ctl call $ACTOR_ID HandleRequest '{"method":"GET", "body":"", "queryString":"", "path":"/test", "header":{}}'
# stop actor across all hosts
$ wash ctl get hosts -o json | jq -r '.hosts[].id' | xargs -n 1 -I{} wash ctl stop actor {} $ACTOR_ID
# show that the actor is reported as stopped across all hosts
$ get-inventory
# Actor should be gone, yet the following call succeeds:
$ wash ctl call $ACTOR_ID HandleRequest '{"method":"GET", "body":"", "queryString":"", "path":"/should_be_dead", "header":{}}'
# Final inventory to show that the actor hasn't been automatically started again by wash.
$ get-inventory
```

**Example output**
```bash
jsoverson@pop-os: ~/development/src/scratch
$ alias get-inventory="wash ctl get hosts -o json | jq -r '.hosts[].id' | xargs -n 1 wash ctl get inventory -o json | jq -s 'map({host:.inventory.host_id, actors: (.inventory.actors)})'"
 
jsoverson@pop-os: ~/development/src/scratch
$ get-inventory                                                                                                                              [14:32:08]
[
  {
    "host": "NDCNHJL32ZD6MNK5Y4WBJ7BWZMX54B5LAJQOCU3AU6T5U6TTQJOLZMHC",
    "actors": []
  }
]
 
jsoverson@pop-os: ~/development/src/scratch
$ wash ctl start actor wasmcloud.azurecr.io/echo:0.2.0                                                                                       [14:32:13]
⡋⠁  Starting actor wasmcloud.azurecr.io/echo:0.2.0 ... 
Actor starting on host NDCNHJL32ZD6MNK5Y4WBJ7BWZMX54B5LAJQOCU3AU6T5U6TTQJOLZMHC
 
jsoverson@pop-os: ~/development/src/scratch
$ get-inventory                                                                                                                              [14:32:16]
[
  {
    "host": "NDCNHJL32ZD6MNK5Y4WBJ7BWZMX54B5LAJQOCU3AU6T5U6TTQJOLZMHC",
    "actors": [
      {
        "id": "MBCFOPM6JW2APJLXJD3Z5O4CN7CPYJ2B4FTKLJUR5YR5MITIU7HD3WD5",
        "image_ref": "wasmcloud.azurecr.io/echo:0.2.0"
      }
    ]
  }
]
 
jsoverson@pop-os: ~/development/src/scratch
$ ACTOR_ID=$(wash claims inspect -o json wasmcloud.azurecr.io/echo:0.2.0 | jq -r '.module')                                                  [14:32:19]
 
jsoverson@pop-os: ~/development/src/scratch
$ wash ctl call $ACTOR_ID HandleRequest '{"method":"GET", "body":"", "queryString":"", "path":"/test", "header":{}}'                         [14:32:27]
⡃⠀ Calling actor MBCFOPM6JW2APJLXJD3Z5O4CN7CPYJ2B4FTKLJUR5YR5MITIU7HD3WD5 ... 
Call response (raw): ��statusCode�Ȧstatus�OK�header��body�H{"method":"GET","path":"/test","query_string":"","headers":{},"body":[]}
 
jsoverson@pop-os: ~/development/src/scratch
$ wash ctl get hosts -o json | jq -r '.hosts[].id' | xargs -n 1 -I{} wash ctl stop actor {} $ACTOR_ID                                        [14:32:31]
⠍⠁  Stopping actor MBCFOPM6JW2APJLXJD3Z5O4CN7CPYJ2B4FTKLJUR5YR5MITIU7HD3WD5 ... 
Stopping actor: MBCFOPM6JW2APJLXJD3Z5O4CN7CPYJ2B4FTKLJUR5YR5MITIU7HD3WD5
 
jsoverson@pop-os: ~/development/src/scratch
$ get-inventory                                                                                                                              [14:32:37]
[
  {
    "host": "NDCNHJL32ZD6MNK5Y4WBJ7BWZMX54B5LAJQOCU3AU6T5U6TTQJOLZMHC",
    "actors": []
  }
]
 
jsoverson@pop-os: ~/development/src/scratch
$ wash ctl call $ACTOR_ID HandleRequest '{"method":"GET", "body":"", "queryString":"", "path":"/should_be_dead", "header":{}}'               [14:32:43]
⢃⠀ Calling actor MBCFOPM6JW2APJLXJD3Z5O4CN7CPYJ2B4FTKLJUR5YR5MITIU7HD3WD5 ... 
Call response (raw): ��statusCode�Ȧstatus�OK�header��body�R{"method":"GET","path":"/should_be_dead","query_string":"","headers":{},"body":[]}
 
jsoverson@pop-os: ~/development/src/scratch
$ get-inventory                                                                                                                              [14:32:45]
[
  {
    "host": "NDCNHJL32ZD6MNK5Y4WBJ7BWZMX54B5LAJQOCU3AU6T5U6TTQJOLZMHC",
    "actors": []
  }
]
```

**Expected behavior**
The second call to the actor should fail with deadline exceeded. This also prevents actors from being properly updated so updates should also succeed.

**Environment (please complete the following information) **
- OS: Linux, ubuntu & popos 
- Architecture: x86_64
- Shell: zsh
- wasmcloud version: 0.15.0

**Comment by autodidaddict** [comment]
I think I may have an idea what this is. In actix, when all references to an actor are gone (dropped), then the actor stops. There's a line of code here: 

https://github.com/wasmCloud/wasmCloud/blob/87e5e0dd64b4bafe2710af42066ec629bb4c071f/crates/wasmcloud-host/src/host_controller/hc_actor.rs#L302

this basically assumes that by removing the actor from that one actors collection that all references to that actor are gone. I suspect that somewhere during the 0.15 work I added one or more additional areas that maintain a reference to that actor, making it so that even after it gets purged from that collection it doesn't "go away"

I'd be curious to see if you get the trace output `trace!("Stopping actor {} per request.", msg.actor_ref);` when calling stop actor.

**Comment by autodidaddict** [comment]
I've just written an integration test that verifies that actors do indeed get shut down when you invoke `stop_actor`:

```rust
pub async fn start_and_stop_actor() -> Result<()> {
    ::std::env::remove_var("KVCACHE_NATS_URL");
    let h = HostBuilder::new().build();
    h.start().await?;
    let echo = Actor::from_file("./tests/modules/echo.wasm")?;
    let actor_id = echo.public_key();
    h.start_actor(echo).await?;
    await_actor_count(&h, 1, Duration::from_millis(50), 3).await?;

    h.stop_actor(&actor_id).await?;
    delay_for(Duration::from_millis(500)).await;
    assert_eq!(0, h.get_actors().await?.len());

    let request = Request {
        method: "GET".to_string(),
        path: "/foo/bar".to_string(),
        query_string: "test=kthxbye".to_string(),
        header: Default::default(),
        body: b"This is a test. Do not be alarmed".to_vec(),
    };
    let buf = serialize(&request)?;
    println!("{}", buf.len());
    let res = h.call_actor(&actor_id, "HandleRequest", &buf).await;
    assert!(true, res.is_err()); // We should not still be able to call this actor

    Ok(())
}
```
This test is passing regularly and predictably. When the control interface receives a "Stop actor" command, it sends the same message to the host controller that the `stop_actor` function does, so the internal mechanisms are identical.

@brooksmtownsend  - are you checking the value of the `failure` field on the ack that wash gets back from the control interface?

**Comment by autodidaddict** [comment]
Some progress. I think I've identified the source of this problem. When a host starts an actor without an RPC connection, it simply starts the actor. Under these conditions, we can issue a stop actor command from either the host api or the lattice control interface and the actor will stop. We now have an integration test to ensure this regression doesn't come up.

When we start an actor with an RPC connection (e.g. in `multiplayer`/`lattice` mode), we create an `RpcSubscription` proxy actor that listens to the NATS topic for RPC calls and then forwards them to the contained actor. When we remove interest from the message bus, I assumed that dropping the message bus's reference to the RPC subscription would be enough... but apparently this doesn't make the proxy actor stop, and even after the actor is removed from the host inventory (the host controller's `actors` collection), the proxy remains.

I've got a failing test in the integration suite now and I'm going to try and refactor the system so that terminating an actix actor for a provider or actor explicitly invokes `ctx.stop()` and the `RpcSubscription` actor forwards the halt invocation to its internal target. This _should_ ensure that the actors stop.
